tests/test_edge_cases.py

```diff
@@ -1506,3 +1524,108 @@ class Model(BaseModel):
+def test_custom_generic_validators():
+    T1 = TypeVar('T1')
+    T2 = TypeVar('T2')
+
+    class MyGen(Generic[T1, T2]):
+        def __init__(self, t1: T1, t2: T2):
+            self.t1 = t1
+            self.t2 = t2
+
+        @classmethod
+        def __get_validators__(cls):
+            yield cls.validate
+
+        @classmethod
+        def validate(cls, v, field):
+            if not isinstance(v, cls):
+                raise ValueError('Invalid value')
+            errors = []
+            for f, name in zip(field.sub_fields, ['t1', 't2']):
+                current_value = getattr(v, name)
+                value, error = f.validate(current_value, {}, loc=name)
+                if error:
+                    errors.append(error)
+            if errors:
+                raise ValidationError(errors, cls)
+            return v
+
+    class Model(BaseModel):
+        a: str
+        gen: MyGen[str, bool]
+
+    with pytest.raises(ValidationError) as exc_info:
+        Model(a='foo', gen='invalid')
+    assert exc_info.value.errors() == [{'loc': ('gen',), 'msg': 'Invalid value', 'type': 'value_error'}]
+
+    with pytest.raises(ValidationError) as exc_info:
+        Model(a='foo', gen=MyGen(t1='bar', t2='baz'))
+    assert exc_info.value.errors() == [
+        {'loc': ('gen', 't2'), 'msg': 'value could not be parsed to a boolean', 'type': 'type_error.bool'}
+    ]
+
+    m = Model(a='foo', gen=MyGen(t1='bar', t2=True))
+    assert m.a == 'foo'
+    assert m.gen.t1 == 'bar'
+    assert m.gen.t2 is True
+
+
+def test_custom_generic_arbitrary_allowed():
+    T1 = TypeVar('T1')
+    T2 = TypeVar('T2')
+
+    class MyGen(Generic[T1, T2]):
+        def __init__(self, t1: T1, t2: T2):
+            self.t1 = t1
+            self.t2 = t2
+
+    class Model(BaseModel):
+        a: str
+        gen: MyGen[str, bool]
+
+        class Config:
+            arbitrary_types_allowed = True
+
+    with pytest.raises(ValidationError) as exc_info:
+        Model(a='foo', gen='invalid')
+    assert exc_info.value.errors() == [
+        {
+            'loc': ('gen',),
+            'msg': 'instance of MyGen expected',
+            'type': 'type_error.arbitrary_type',
+            'ctx': {'expected_arbitrary_type': 'MyGen'},
+        }
+    ]
+
+    # No validation, no exception
+    m = Model(a='foo', gen=MyGen(t1='bar', t2='baz'))
+    assert m.a == 'foo'
+    assert m.gen.t1 == 'bar'
+    assert m.gen.t2 == 'baz'
+
+    m = Model(a='foo', gen=MyGen(t1='bar', t2=True))
+    assert m.a == 'foo'
+    assert m.gen.t1 == 'bar'
+    assert m.gen.t2 is True
+
+
+def test_custom_generic_disallowed():
+    T1 = TypeVar('T1')
+    T2 = TypeVar('T2')
+
+    class MyGen(Generic[T1, T2]):
+        def __init__(self, t1: T1, t2: T2):
+            self.t1 = t1
+            self.t2 = t2
+
+    with pytest.raises(TypeError) as exc_info:
+
+        class Model(BaseModel):
+            a: str
+            gen: MyGen[str, bool]
+
+    assert 'Fields of type' in str(exc_info.value)
```

Comment:
better to use `pytest.raises(TypeError, match='...')`

Reply:
I didn't know about `pytest.raises(TypeError, match='...')` :scream: